media/js/base/mozilla-client.js

```diff
@@ -0,0 +1,171 @@
+            if (!strict) {
+                latestVerArr.length = 1;
+            }
+
+            for (var j = 0; j < userVerArr.length; j++) {
+                if (Number(userVerArr[j]) < Number(latestVerArr[j] || 0)) {
+                    isUpToDate = false;
+                    break;
+                } else {
+                    isUpToDate = true;
+                }
+            }
+
+            if (isUpToDate) {
+                break;
+            }
+        }
+
+        return isUpToDate;
+    };
+
+    /**
+     * Use the async mozUITour API of Firefox to retrieve the user's browser info, including the update channel and
+     * accurate, patch-level version number. This API is available on Firefox 35 and later. See
+     * http://bedrock.readthedocs.org/en/latest/uitour.html for details.
+     *
+     * @param  {Function} callback - callback function to be executed with the Firefox details
+     * @return {None}
+     */
+    Client.getFirefoxDetails = function (callback) {
+        var self = this;
+        var ssKeyName = 'Mozilla.Client.FirefoxDetails';
+        var callbackID = Math.random().toString(36).replace(/[^a-z]+/g, '');
+
+        // Try to find cache in sessionStorage first. Use try-catch in case sessionStorage is unimplemented or disabled
+        try {
+            var cache = sessionStorage.getItem(ssKeyName);
+
+            if (cache && !this.FirefoxDetails) {
+                this.FirefoxDetails = JSON.parse(cache);
+            }
+        } catch (ex) {}
+
+        // Fire the callback function immediately if cache exists
+        if (this.FirefoxDetails) {
+            callback(this.FirefoxDetails);
+
+            return;
+        }
+
+        var listener = function (event) {
+            if (!event.detail || !event.detail.data || event.detail.callbackID !== callbackID) {
+                return;
+            }
+
+            window.clearTimeout(fallback);
+            onRetrieved(true, event.detail.data.version, event.detail.data.defaultUpdateChannel);
+        };
+
+        // Prepare fallback function in case the API doesn't work. This will be fired in .2 seconds
+        var fallback = window.setTimeout(function () {
+            onRetrieved(false, self._getFirefoxVersion(), 'release');
+        }, 200);
+
+        var onRetrieved = function (accurate, version, channel) {
+            document.removeEventListener('mozUITourResponse', listener);
+
+            var isESR = channel === 'esr';
+            var isUpToDate = self._isFirefoxUpToDate(accurate, accurate ? isESR : false, version);
+            var details = self.FirefoxDetails = {
+                'version': version,
+                'channel': channel,
+                'isUpToDate': isUpToDate,
+                'isESR': isESR
+            };
+
+            callback(details);
+
+            // Cache the data with sessionStorage
+            try {
+                sessionStorage.setItem(ssKeyName, JSON.stringify(details));
```

Comment:
Like I mentioned in the previous comment, I don't think we should be storing user data is session storage like this unless there is a really good reason. Session storage is also synchronous, so another reason I think we should avoid using it.